tests/test_project.py

```diff
@@ -1787,6 +1788,154 @@ class TestGroupDynamicProjectMainInterface(TestProjectMainInterface):
+        assert agg_cursor._doc_filter is None
+        # All operations will return aggregates, even if the aggregates are not
+        # unique to that operation, because every operation/group in this
+        # project has a custom aggregator defined. Only the default aggregator
+        # can de-duplicate the returned results in the cursor, thus it is
+        # expected that the length of the cursor is larger than the number of
+        # unique ids present in the cursor.
+        assert len(agg_cursor) == 41
+        assert len({get_aggregate_id(agg) for agg in agg_cursor}) == 34
+        assert tuple(project) in agg_cursor
+        assert all((job,) in agg_cursor for job in project)
+
+    def test_filters(self):
+        project = self.mock_project()
+        agg_cursor = _AggregatesCursor(project=project, filter={"even": True})
+        assert agg_cursor._project == project
+        assert agg_cursor._filter == {"even": True}
+        assert agg_cursor._doc_filter is None
+        assert len(agg_cursor) == 15
+
+    def test_reregister_aggregates(self):
+        project = self.mock_project()
+        agg_cursor = _AggregatesCursor(project=project)
+        assert len(agg_cursor) == 41
+        project.open_job(dict(i=31, even=False)).init()
+        # Default aggregate store doesn't need to be re-registered.
+        assert len(agg_cursor) == 42
+        project.reregister_aggregates()
+        # The operation agg_op2 adds another aggregate in the project.
+        assert len(agg_cursor) == 43
+
+
+class TestAggregationProjectMainInterface(TestAggregatesProjectBase):
+    def test_main_run(self):
+        project = self.mock_project()
+        assert len(project)
+        for job in project:
+            assert not job.doc.get("sum", False)
+            assert not job.doc.get("sum_other", False)
+            assert not job.doc.get("sum_custom", False)
+            assert not job.doc.get("op2", False)
+            assert not job.doc.get("op3", False)
+
+        even_sum = sum([job.sp.i for job in project if job.sp.i % 2 == 0])
+        odd_sum = sum([job.sp.i for job in project if job.sp.i % 2 != 0])
+
+        self.call_subcmd(
+            "run -o agg_op1 agg_op1_different agg_op1_custom agg_op2 agg_op3 --show-traceback"
+        )
+
+        for job in project:
+            assert job.doc.op2
+            assert job.doc.op3
+            if job.sp.i % 2 == 0:
+                assert (
+                    job.doc.sum == job.doc.sum_other == job.doc.sum_custom == even_sum
+                )
+            else:
+                assert job.doc.sum == job.doc.sum_other == job.doc.sum_custom == odd_sum
+
+    def test_invalid_with_job(self):
+        project = self.mock_project()
+        assert len(project)
+
+        with pytest.raises(Exception) as runinfo:
+            project.run(names=["agg_op5"])
+            assert (
+                "use of @with_job decorator with aggregation" in runinfo.value.message
+            )
```

Comment:
Can we not detect this earlier than trying to run the operation?

Reply:
Fixed in 865277b2a7b65c31bf2fe9b8ac0a183b073cd5b1.